File: src/main/java/io/github/hiiragi283/api/shape/HTShapeRegistry.java
package io.github.hiiragi283.api.shape;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import com.google.common.collect.ImmutableMap;

public final class HTShapeRegistry {

    public final ImmutableMap<HTShapeKey, HTShape> registry;

    public HTShapeRegistry(Map<HTShapeKey, HTShape> map) {
        registry = ImmutableMap.copyOf(map);
    }

    @NotNull
    public Set<HTShapeKey> getKeys() {
        return registry.keySet();
    }

    @NotNull
    public Collection<HTShape> getValues() {
        return registry.values();
    }

    @Nullable
    public HTShape getShape(@NotNull HTShapeKey shapeKey) {
        return registry.get(shapeKey);
    }

    public boolean hasShape(@NotNull HTShapeKey shapeKey) {
        return registry.containsKey(shapeKey);
    }
}

File: src/main/kotlin/io/github/hiiragi283/bwms/api/shape/HTShapeRegistry.kt
package io.github.hiiragi283.bwms.api.shape

import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import com.google.common.collect.ImmutableBiMap

class HTShapeRegistry private constructor(private val biMap: BiMap<HTShapeKey, HTShape>) {
	companion object {
		@JvmField
		val EMPTY = HTShapeRegistry(ImmutableBiMap.of())

		@JvmStatic
		fun create(builderAction: Builder.() -> Unit): HTShapeRegistry {
			val map: MutableMap<HTShapeKey, HTShapeBuilder> = mutableMapOf()
			Builder(map).apply(builderAction)
			return map
				.mapValues { it.value.build() }
				.let { HashBiMap.create(it) }
				.let(::HTShapeRegistry)
		}
	}

	//    Map    //

	val keys: Set<HTShapeKey>
		get() = biMap.keys

	val shapes: Collection<HTShape>
		get() = biMap.values

	operator fun contains(key: HTShapeKey): Boolean = key in biMap

	operator fun get(key: HTShapeKey): HTShape? = biMap[key]

	fun getOrDefault(key: HTShapeKey, defaultValue: HTShape): HTShape = get(key) ?: defaultValue

	fun getKey(shape: HTShape): HTShapeKey? = biMap.inverse()[shape]

	fun forEach(action: (HTShapeKey, HTShape) -> Unit) {
		biMap.forEach(action)
	}

	//    Builder    //

	class Builder(private val map: MutableMap<HTShapeKey, HTShapeBuilder>) {
		fun createBlockShape(key: HTShapeKey): HTShapeBuilder {
			check(key !in map) { "Shape builder; $key is already created!" }
			return map
				.computeIfAbsent(key) { HTShapeBuilder.createBlock(key) }
				.apply { key.validated = true }
		}

		fun createItemShape(key: HTShapeKey): HTShapeBuilder {
			check(key !in map) { "Shape builder; $key is already created!" }
			return map
				.computeIfAbsent(key) { HTShapeBuilder.createItem(key) }
				.apply { key.validated = true }
		}

		fun getBuilder(key: HTShapeKey): HTShapeBuilder? = map[key]

		fun canGenerateItem(key: HTShapeKey): Boolean = getBuilder(key)?.allowItemGen == true
	}
}
